File: server/lib/pdfSplit.ts
/**
 * lib/pdfSplit.ts — page-count + page-range splitting for the W1 native-PDF path.
 *
 * When a study report is dense enough to risk the model's output-token ceiling
 * in a single native-PDF call, extractArcFlashDocument cuts it into OVERLAPPING
 * page windows (so a table straddling a seam is whole in at least one window)
 * and sends each window's sub-PDF natively, merging the results by bus name.
 * This module produces those sub-PDF buffers via scripts/pdf_split.py
 * (pypdfium2, already in the image) — same spawn-python pattern as rasterizePdf.
 *
 * BEST-EFFORT by design: any failure returns 0 pages / [] so the caller falls
 * back cleanly to a single call or the deterministic text/vision path.
 */

'use strict';

const { execFile } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');
const crypto = require('crypto');

const PY = process.env.PYEXTRACT_PYTHON || 'python3';
const SCRIPT = path.join(__dirname, '..', 'scripts', 'pdf_split.py');
const TIMEOUT_MS = parseInt(process.env.PYEXTRACT_TIMEOUT_MS || '45000', 10);

function runSplit(args: string[], timeout: number): Promise<any> {
  return new Promise((resolve) => {
    execFile(PY, [SCRIPT, ...args], { timeout, maxBuffer: 16 * 1024 * 1024 }, (err: any, stdout: any) => {
      if (err || !stdout) return resolve({ ok: false });
      try {
        const out = JSON.parse(String(stdout).trim().split('\n').pop());
        resolve(out && out.ok ? out : { ok: false });
      } catch {
        resolve({ ok: false });
      }
    });
  });
}

// True total page count of the PDF (0 on any failure). Used to decide whether a
// single native call is safe or the document must be windowed.
async function pdfPageCount(buffer: Buffer): Promise<number> {
  let tmp: string | null = null;
  try {
    tmp = path.join(os.tmpdir(), `afsplit-${crypto.randomBytes(8).toString('hex')}.pdf`);
    fs.writeFileSync(tmp, buffer);
  } catch {
    return 0;
  }
  try {
    const out = await runSplit([tmp, 'count'], TIMEOUT_MS);
    return out && out.ok ? (out.pages || 0) : 0;
  } finally {
    try { if (tmp) fs.unlinkSync(tmp); } catch { /* tmpfs reaps anyway */ }
  }
}

// Split into sub-PDF buffers, one per [startPage, endPage] range (1-based,
// inclusive). Returns [] on any failure so the caller can fall back.
async function splitPdfByRanges(buffer: Buffer, ranges: Array<[number, number]>): Promise<Buffer[]> {
  if (!Array.isArray(ranges) || !ranges.length) return [];
  let dir: string | null = null;
  try {
    dir = fs.mkdtempSync(path.join(os.tmpdir(), 'afsplit-'));
  } catch {
    return [];
  }
  const inPath = path.join(dir, 'in.pdf');
  const prefix = path.join(dir, 'part');
  try {
    fs.writeFileSync(inPath, buffer);
    const rangeStr = ranges.map(([a, b]) => `${a}-${b}`).join(',');
    const out = await runSplit([inPath, 'split', prefix, rangeStr], TIMEOUT_MS);
    if (!out || !out.ok || !Array.isArray(out.files)) return [];
    const buffers: Buffer[] = [];
    for (const f of out.files) {
      try { buffers.push(fs.readFileSync(f)); } catch { /* skip missing part */ }
    }
    return buffers;
  } catch {
    return [];
  } finally {
    try { if (dir) fs.rmSync(dir, { recursive: true, force: true }); } catch { /* best effort */ }
  }
}

export { pdfPageCount, splitPdfByRanges };

File: server/scripts/pdf_split.py
#!/usr/bin/env python3
"""
W1 native-PDF splitter via pypdfium2 (already in the image; NO AI tokens).

The native-PDF arc-flash path sends the whole PDF to the model in one call. When
a report is dense enough to risk the model's output-token ceiling, the Node
caller cuts it into OVERLAPPING page windows and sends each window's sub-PDF
natively, then merges the extractions by bus name. Overlap guarantees a table
that straddles a window seam is whole in at least one window, so no bus is lost
or split at a chunk boundary.

Best-effort: prints one JSON line; {"ok": false} on any failure so ingest can
always fall back to the deterministic text/vision path.

Usage:
  pdf_split.py <in.pdf> count
  pdf_split.py <in.pdf> split <out_prefix> <ranges>
      ranges = "1-2,2-3,3-4"  (1-based, inclusive page ranges)
      writes <out_prefix>-0.pdf, <out_prefix>-1.pdf, ... in range order
"""
import sys
import json


def main():
    if len(sys.argv) < 3:
        print(json.dumps({"ok": False}))
        return 2
    pdf_path = sys.argv[1]
    mode = sys.argv[2]

    import pypdfium2 as pdfium  # imported here so an import error fails open below

    src = pdfium.PdfDocument(pdf_path)
    pages = len(src)

    if mode == "count":
        print(json.dumps({"ok": True, "pages": pages}))
        return 0

    if mode == "split":
        if len(sys.argv) < 5:
            print(json.dumps({"ok": False}))
            return 2
        out_prefix = sys.argv[3]
        ranges = sys.argv[4]
        files = []
        for idx, part in enumerate(r for r in ranges.split(",") if r.strip()):
            a_s, b_s = part.split("-")
            a = max(1, int(a_s))
            b = min(pages, int(b_s))
            if b < a:
                continue
            dst = pdfium.PdfDocument.new()
            dst.import_pages(src, pages=list(range(a - 1, b)))
            out = "%s-%d.pdf" % (out_prefix, idx)
            dst.save(out)
            dst.close()
            files.append(out)
        print(json.dumps({"ok": True, "pages": pages, "files": files}))
        return 0

    print(json.dumps({"ok": False}))
    return 2


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as exc:  # noqa: BLE001 — any failure is a clean fail-open
        sys.stderr.write("pdf_split failed: %s\n" % exc)
        print(json.dumps({"ok": False}))
        sys.exit(1)
